src/Chainweb/CutDB.hs

```diff
@@ -590,24 +591,28 @@ cutHashesToBlockHeaderMap
     -> IO (Either (HM.HashMap ChainId BlockHash) (HM.HashMap ChainId BlockHeader))
         -- ^ The 'Left' value holds missing hashes, the 'Right' value holds
         -- a 'Cut'.
-cutHashesToBlockHeaderMap logfun headerStore payloadStore hs =
-    trace logfun "Chainweb.CutDB.cutHashesToBlockHeaderMap" (_cutId hs) 1 $ do
-        plds <- emptyCas
-        casInsertBatch plds $ V.fromList $ HM.elems $ _cutHashesPayloads hs
-
-        hdrs <- emptyCas
-        casInsertBatch hdrs $ V.fromList $ HM.elems $ _cutHashesHeaders hs
-
-        (headers :> missing) <- S.each (HM.toList $ _cutHashes hs)
-            & S.map (fmap snd)
-            & S.mapM (tryGetBlockHeader hdrs plds)
-            & S.partitionEithers
-            & S.fold_ (\x (cid, h) -> HM.insert cid h x) mempty id
-            & S.fold (\x (cid, h) -> HM.insert cid h x) mempty id
-        if null missing
-            then return $! Right headers
-            else return $! Left missing
+cutHashesToBlockHeaderMap logfun headerStore payloadStore hs = timeout 1000000 go >>= \case
+    Nothing -> return $! Left mempty
+    Just x -> return $! x
   where
+    go =
+        trace logfun "Chainweb.CutDB.cutHashesToBlockHeaderMap" (_cutId hs) 1 $ do
+            plds <- emptyCas
+            casInsertBatch plds $ V.fromList $ HM.elems $ _cutHashesPayloads hs
+
+            hdrs <- emptyCas
+            casInsertBatch hdrs $ V.fromList $ HM.elems $ _cutHashesHeaders hs
+
+            (headers :> missing) <- S.each (HM.toList $ _cutHashes hs)
+                & S.map (fmap snd)
+                & S.mapM (tryGetBlockHeader hdrs plds)
+                & S.partitionEithers
+                & S.fold_ (\x (cid, h) -> HM.insert cid h x) mempty id
+                & S.fold (\x (cid, h) -> HM.insert cid h x) mempty id
```

Comment:
why these back-to-back folds?

Reply:
this is unrelated to the PR. Let's take this offline.